### PR TITLE
Updated review app (templates, views, etc.)

### LINKED ISSUE #80: Add review templates
under `vagrant/review/templates/review/*.html`

### LINKED ISSUE #81: Add review views.py
under `vagrant/review/templates/review/views.py`

### LINKED ISSUE #82: Add review URLs
under `vagrant/Pro4U/urls.py`

### LINKED ISSUE #92: Update ReviewManager functions
All functions sort **all reviews** ascending/descending by date or rating
But we need to sort filtered reviews by professional id 
not to sort all the reviews of all professionals

### PR DESCRIPTION
**Pull request by guybeckenstein**
Resolves #92 
Resolves #82 
Resolves #81 
Resolves #80 
#
review form
review list

# `review/models.py`
All sort functions now get **professional id** as parameter when used

# `review/tests/test_review_manager.py`
It was needed to update the tests within test class, because they failed after update

# Views
Created three classes. 
1. ReviewListView functionality:
   - Pagination of all reviews, to professional with ID=pk.
   - Sorting reviews by date (oldest/newest) or rating (lowest/highest)
   - Searching reviews by keywords (of the description only)
2. ReviewCreateView functionality:
   - Using ReviewForm from `reviews/forms.py`.
   - Needs only rating (from 1 to 5, or undefined) and description. Description is not mandatory, rating is.
   - If user X has already reviewed professional Y, he will be redirected to ReviewUpdateView
3. ReviewUpdateView functionality:
   - Same logic as ReviewUpdateView, except redirection.
Testing was currently added only to `ReviewListView`. I need help with testing other views.
# Templates
- Created two templates (except base template).
- Used bootstrap instead of CSS (ReviewCreateView, ReviewUpdateView)
1. `review/review_list.html` functionality:
   - Inherits from `review/review_base.html`
   - If professional with ID=pk has reviews:
       - Shows average rating (if at least one review rating is defined)
       - Counts the reviews
       - Allows search functionality
       - Allows sorting functionality
   - Else:
       -  Blank page with a text that says that professional with ID=pk has no reviews.
   - Allows user without a prior review to professional with ID=pk to write a new review (using blue button).
   - If user already reviewed, allows it to edit the review.
2. `review/review_form.html` functionality:
   - Inherits from `review/review_base.html`
   - Once the review **CREATED** using CREATE method, it's date is forever (no matter how many edits user will do)
   - Submit only if a rating was given
Testing was currently added only to `review/review_list.html`. I need help with testing other templates.

# JavaScript files
For allowing the **infinite scrolldown** instead of normal pagination, I adjusted the code, and used JS scripts.
The scripts were written by someone from youtube

# Configurations
Updated `urls.py`

### CODE REVIEW

**Review by Yarboa — CHANGES_REQUESTED** [review]
Please remove the usage of client, client2 fixtures
You can use this instead

https://github.com/redhat-beyond/Pro4U/blob/main/conftest.py#LL114

File: review/tests/conftest.py [hunk 1]
```diff
@@ -9,15 +9,15 @@
 
 @pytest.fixture(scope='function')
 def create_review() -> Callable[[str, str, str, str, int], Review]:
-    def _review_factory(client: str, professional: str, rating: str, description: str, days: int) -> Review:
+    def _review_factory(demo_client: str, professional: str, rating: str, description: str, days: int) -> Review:
```
Comment (Yarboa) [inline review/tests/conftest.py:12]: Please use the make_client fixture instead of client
https://github.com/redhat-beyond/Pro4U/blob/main/conftest.py#L114

You can override parameters if needed

same goes for make_professional fixture

File: review/tests/test_review_manager.py [hunk 2]
```diff
@@ -39,25 +39,25 @@ def test_sort_review_by_lowest_rating(self, demo_client, demo_client2, professio
...
-                           professional_full_name='Tal Reinfeld')
+        sorted_reviews = Review.objects.sort_review_by_lowest_rating(professional.professional_id)
+        self.assert_review(sorted_reviews.first(), rating='2',
+                           client_full_name='Bob Builder',
+                           professional_full_name='Bob Builder')
+        self.assert_review(sorted_reviews.last(), rating=Review.Rating.THREE_STARS,
+                           client_full_name='Bob Builder',
+                           professional_full_name='Bob Builder')
 
     def test_sort_review_by_highest_rating(self, demo_client, demo_client2, professional,
```
Comment (Yarboa) [inline review/tests/test_review_manager.py:50]: same as above
make_client, make_professional fixtures

File: review/tests/test_review_manager.py [hunk 3]
```diff
@@ -39,25 +39,25 @@ def test_sort_review_by_lowest_rating(self, demo_client, demo_client2, professio
                                           create_review: Callable[[str, str, str, str, int], Review]):
         self.create_reviews(demo_client, demo_client2, professional, create_review)
```
Comment (Yarboa) [inline review/tests/test_review_manager.py:40]: same as above

**Review by Yarboa — COMMENTED** [review]
It is wise to have less capabilities in each PR

File: review/tests/conftest.py (hunk 1 above)
Comment (guybeckenstein) [reply review/tests/conftest.py:12]: > Please use the make_client fixture instead of client https://github.com/redhat-beyond/Pro4U/blob/main/conftest.py#L114
> 
> You can override parameters if needed
> 
> same goes for make_professional fixture

Both fixtures still exist, should I change?

File: review/tests/test_review_create.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+# from django.urls import reverse
+#
+#
+# @pytest.mark.django_db
+# class TestCreate:
+#     def test_review_create(self, client, professional):
+#         url = reverse('review-create', kwargs={'pk': professional.pk})
+#         response = client.get(url)  # This breaks the code
+#         assert response.status_code == 200
+#         assert 'review/review_form.html' in response.templates[0].name
```
Comment (idosinger98) [inline review/tests/test_review_create.py:11]: Please delete it if you don't need it

File: Pro4U/urls.py [hunk 5]
```diff
@@ -19,9 +19,15 @@
...
+from review import views as review_views
+from review.models import Review
 
 urlpatterns = [
     path('admin/', admin.site.urls),
     path('', views.indexView, name='invokeIndexPage'),
+    path('professional/<int:pk>/reviews/', review_views.ReviewListView.as_view(model=Review, paginate_by=10),
+         name='reviews'),
+    path('professional/<int:pk>/reviews/new/', review_views.ReviewCreateView.as_view(), name='review-create'),
+    path('professional/<int:pk>/reviews/update/', review_views.ReviewUpdateView.as_view(), name='review-update'),
```
Comment (idosinger98) [inline Pro4U/urls.py:31]: What about deleting a review?

File: review/tests/test_review_update.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+# from django.urls import reverse
+#
+#
+# @pytest.mark.django_db
+# class TestUpdate:
+#     def test_review_update(self, client, professional):
+#         url = reverse('review-update', kwargs={'pk': professional.pk})
+#         response = client.get(url)  # This breaks the code
+#         assert response.status_code == 200
+#         assert 'review/review_form.html' in response.templates[0].name
```
Comment (idosinger98) [inline review/tests/test_review_update.py:11]: Why do you need this file if everything is here in the comments?

File: templates/base_reservation.html [hunk 7]
```diff
@@ -1,4 +1,4 @@
-<!-- cal/templates/cal/base.html -->
+<!-- cal/templates/cal/base_review.html -->
```
Comment (idosinger98) [inline templates/base_reservation.html:1]: Why did you change it?

File: review/views.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    #
+    # def get_context_data(self, **kwargs):
+    #     # This enables to see what is the name of the professional in the HTML template title
+    #     context = super().get_context_data(**kwargs)
+    #     context['professional'] = Professional.objects.get(professional_id=self.kwargs['pk'])
+    #     return context
+    #
+    # def get_success_url(self):
+    #     return reverse('reviews', args=[self.kwargs['pk']])
+
```
Comment (idosinger98) [inline review/views.py:133]: Please delete it if you don't need it

File: review/views.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    # def get_success_url(self):
+    #     return reverse('reviews', args=[self.kwargs['pk']])
+    #
+    # def test_func(self):
+    #     # Only client X can update reviews of client X
+    #     review = self.get_object()
+    #     profile = self.request.user.profile
+    #     client, _ = Client.objects.get_or_create(profile_id=profile)
+    #     return client == review.client
```
Comment (idosinger98) [inline review/views.py:183]: Please delete it if you don't need it

File: review/tests/test_review_create.py (hunk 4 above)
Comment (guybeckenstein) [reply review/tests/test_review_create.py:11]: I will fix it

File: Pro4U/urls.py (hunk 5 above)
Comment (guybeckenstein) [reply Pro4U/urls.py:31]: no need to delete a review

File: review/tests/test_review_update.py (hunk 6 above)
Comment (guybeckenstein) [reply review/tests/test_review_update.py:11]: I will fix it

File: templates/base_reservation.html (hunk 7 above)
Comment (guybeckenstein) [reply templates/base_reservation.html:1]: I will fix it

File: review/views.py (hunk 8 above)
Comment (guybeckenstein) [reply review/views.py:133]: I will fix it

File: review/views.py (hunk 9 above)
Comment (guybeckenstein) [reply review/views.py:183]: I will fix it

File: review/tests/conftest.py (hunk 1 above)
Comment (Yarboa) [reply review/tests/conftest.py:12]: Correct but @idosinger98 and @TalReinfeld 
Made some changes in the past, and they worked around the demo_client usage by others.
I think they had to delete it, 
Please open a ticket for removing the fixtures that do not use the factories

**Review by idosinger98 — APPROVED** [review]
LGTM, Resolve conflicts

**Comment by guybeckenstein** [comment]
@Yarboa This PR is under construction

File: review/tests/conftest.py (hunk 1 above)
Comment (guybeckenstein) [reply review/tests/conftest.py:12]: > 

I will be using `demo_client`

**Comment by guybeckenstein** [comment]
@Yarboa I finished the PR.

File: review/tests/conftest.py (hunk 1 above)
Comment (Yarboa) [reply review/tests/conftest.py:12]: Please fix it, 
It is working,

Comment (guybeckenstein) [reply review/tests/conftest.py:12]: > Please fix it, It is working,

you mean use `make_client` instead of `demo_client` and `demo_client2`?

Comment (Yarboa) [reply review/tests/conftest.py:12]: Please avoid using demo_client and demo_client2
Those are deprecated methods

Comment (guybeckenstein) [reply review/tests/conftest.py:12]: > Please avoid using demo_client and demo_client2 Those are deprecated methods

Resolved

File: review/tests/test_review_manager.py (hunk 3 above)
Comment (guybeckenstein) [reply review/tests/test_review_manager.py]: > same as above

Resolved (i renamed the file though git thinks i deleted it and created a new one instead)

**Comment by guybeckenstein** [comment]
> Please remove the usage of client, client2 fixtures You can use this instead
> 
> https://github.com/redhat-beyond/Pro4U/blob/main/conftest.py#LL114

Resolved

**Comment by Yarboa** [comment]
> > Please remove the usage of client, client2 fixtures You can use this instead
> > https://github.com/redhat-beyond/Pro4U/blob/main/conftest.py#LL114
> 
> Resolved

https://redhat-beyond.slack.com/archives/C04TLCD78UT/p1685279099534329

**Comment by guybeckenstein** [comment]
> > > Please remove the usage of client, client2 fixtures You can use this instead
> > > https://github.com/redhat-beyond/Pro4U/blob/main/conftest.py#LL114
> > 
> > 
> > Resolved
> 
> https://redhat-beyond.slack.com/archives/C04TLCD78UT/p1685279099534329

Done. Also modified some code in `conftest.py` for more precise python `type(...)`

**Review by idosinger98 — APPROVED** [review]
LGTM

File: review/tests/conftest.py [hunk 10]
```diff
@@ -4,12 +4,13 @@
...
 
+from account.models.client import Client
 from review.models import Review
 
 
 @pytest.fixture(scope='function')
 def create_review() -> Callable[[str, str, str, str, int], Review]:
-    def _review_factory(client: str, professional: str, rating: str, description: str, days: int) -> Review:
+    def _review_factory(client: Client, professional: str, rating: str, description: str, days: int) -> Review:
         # Review function instance, currently without parameters. The main idea is that if `create_review` fixture
```
Comment (Yarboa) [inline review/tests/conftest.py:14]: Please remove

**Review by Yarboa — COMMENTED** [review]
LGTM, Please  remove comments

**Comment by guybeckenstein** [comment]
> LGTM, Please remove comments

done @Yarboa
